### PR TITLE
bpo-1635741: Port audioop extension module to multiphase initialization(PEP 489)

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by shihai1991** [comment]
Looks there have an potential bug, because running test_audioop test case, the error name not belong to audioop: `binascii.Error: Input sample should be longer`

**Comment by encukou** [comment]
I don't have a Windows machine to test this on.
But, I see that the struct name and the macro to get it are both called `_audioopstate`! I missed that in my earlier review :(

Could you first turn the macro into an inline function, and rename the type (to something like `audioop_state`)?

**Comment by shihai1991** [comment]
> I don't have a Windows machine to test this on.
> But, I see that the struct name and the macro to get it are both called `_audioopstate`! I missed that in my earlier review :(
> 
> Could you first turn the macro into an inline function, and rename the type (to something like `audioop_state`)?

of course, let me try it now ;)

**Comment by shihai1991** [comment]
petr, it's failed again, so I try to find a win env to debug~

File: Modules/audioop.c [hunk 1]
```diff
@@ -1903,31 +1923,60 @@ static PyMethodDef audioop_methods[] = {
 };
 
 static int
-audioop_traverse(PyObject *m, visitproc visit, void *arg) {
-    _audioopstate *state = _audioopstate(m);
-    if (state != NULL)
+audioop_traverse(PyObject *m, visitproc visit, void *arg)
+{
+    audioop_state *state = (audioop_state *)PyModule_GetState(m);
+    if (state) {
```
Comment (vstinner) [inline Modules/audioop.c:1929]: Is it really possible that state can be NULL when this function is called?

File: Modules/audioop.c [hunk 2]
```diff
@@ -1903,31 +1923,60 @@ static PyMethodDef audioop_methods[] = {
...
 }
+
 static int
-audioop_clear(PyObject *m) {
-    _audioopstate *state = _audioopstate(m);
-    if (state != NULL)
+audioop_clear(PyObject *m)
+{
+    audioop_state *state = (audioop_state *)PyModule_GetState(m);
+    if (state) {
```
Comment (vstinner) [inline Modules/audioop.c:1939]: Is it really possible that state can be NULL when this function is called?

module_clear() doesn't touch module->md_state value. Only module_dealloc() calls PyMem_FREE(m->md_state). But audioop_clear() must not be called on a deallocated module object.

File: Modules/audioop.c [hunk 3]
```diff
@@ -1903,31 +1923,60 @@ static PyMethodDef audioop_methods[] = {
...
+    if (PyModule_AddObject(m, "error", state->AudioopError) < 0) {
+        Py_DECREF(state->AudioopError);
+        return -1;
+    }
+
+    return 0;
+}
+
+static PyModuleDef_Slot audioop_slots[] = {
+    {Py_mod_exec, audioop_exec}
```
Comment (vstinner) [inline Modules/audioop.c]: I used a debugger on Windows. You forgot the terminator:

```suggestion
    {Py_mod_exec, audioop_exec},
    {0, NULL}
```

Comment (shihai1991) [reply Modules/audioop.c]: Oh, good catch.
Thanks a million, victor.

**Comment by shihai1991** [comment]
Oh, gate have passed, thanks, victor.

File: Modules/audioop.c (hunk 2 above)
Comment (vstinner) [reply Modules/audioop.c:1939]: I would prefer to see this question answered before approving the PR.

Comment (encukou) [reply Modules/audioop.c:1939]: See [the docs](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_clear).

Comment (shihai1991) [reply Modules/audioop.c:1939]: Is it worth us to change this logic as victor said?

Comment (vstinner) [reply Modules/audioop.c:1939]: I created https://bugs.python.org/issue39824 to propose to change the Python behavior, so modules would not have to handle md_state=NULL anymore.

**Pull request merged**